### ISSUE TITLE
reading big file and memory usage

### ISSUE THREAD

**Issue by patatetom** [fusepy/fusepy#96]
hello world,

I have written a [xb360hd module](https://github.com/patatetom/xtaf/blob/master/xb360hd.py) that allows me a read-only access to the contents of an [xtaf](http://www.free60.org/wiki/FATX) partition (xbox360).

now, I'm trying to make this content accessible with fuse, but I'm having a little problem with reading files : a file can be very large (and discontinuous), so I try to avoid storing it in memory using the [yield instruction](https://github.com/patatetom/xtaf/blob/master/xb360hd.py#L209), but I've not been able to use it in the [read function](https://github.com/patatetom/xtaf/blob/master/xtaffuse.py#L52) of the [xtaffuse module](https://github.com/patatetom/xtaf/blob/master/xtaffuse.py).

I would very much appreciate some advice or hints from you on how to carry out my project successfully.

regards, lacsaP.

**Comment by patatetom** [comment]
after a few tests, it appears that the file is read in pieces:

```
read[0] 131072 bytes from 0 flags: 0x8000
read[0] 131072 bytes from 131072 flags: 0x8000
read[0] 131072 bytes from 262144 flags: 0x8000
...
read[0] 8192 bytes from 917504 flags: 0x8000
read[0] 4096 bytes from 925696 flags: 0x8000
```

so the problem I was raising is not a problem in practice.

I would really appreciate your feedback on my [xtaffuse module](https://github.com/patatetom/xtaf/blob/master/xtaffuse.py) :-)

**Comment by terencehonles** [comment]
Is your code working? I'm not sure what you are asking. When comparing to the xb360hd module you shared it appears to behave similarly.

**Comment by patatetom** [comment]
yes, I think so : take a look at the readme of the repository I just updated.

I assumed that I was going to have to read the entire file before returning it to fusepy, which was going to be a problem. but in fact, fusepy is claiming the file in small pieces.

looking at what is returned by the debug mode, I noticed something strange...

if the file is "small" (less than 4096 bytes a priori), the read is "normal" :

```bash
# in the bash console
stat -c %s /tmp/xbox/name.txt 
22
cat /tmp/xbox/name.txt > /dev/null
```
```python
# in the pyhon console
…
unique: 380, opcode: READ (15), nodeid: 2, insize: 80, pid: 2073
read[0] 4096 bytes from 0 flags: 0x8000
   read[0] 4096 bytes from 0
   unique: 380, success, outsize: 4112
…
```

but if the file is "big" (more than 4096 bytes a priori), the read is "strange" :

```bash
stat -c %s /tmp/xbox/Content/E00005538DC276AE/454109C3/00000001/Settings\ 20051122143332 
286720
cat /tmp/xbox/Content/E00005538DC276AE/454109C3/00000001/Settings\ 20051122143332  > /dev/null
```
```python
…
unique: 527, opcode: LOOKUP (1), nodeid: 19, insize: 64, pid: 2101
LOOKUP /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
getattr /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   NODEID: 20
   unique: 527, success, outsize: 144
unique: 528, opcode: OPEN (14), nodeid: 20, insize: 48, pid: 2101
open flags: 0x8000 /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   open[0] flags: 0x8000 /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   unique: 528, success, outsize: 32
unique: 529, opcode: READ (15), nodeid: 20, insize: 80, pid: 2101
read[0] 131072 bytes from 0 flags: 0x8000
   read[0] 131072 bytes from 0
   unique: 529, success, outsize: 131088
unique: 530, opcode: READ (15), nodeid: 20, insize: 80, pid: 2101
read[0] 131072 bytes from 131072 flags: 0x8000
   read[0] 131072 bytes from 131072
   unique: 530, success, outsize: 131088
unique: 531, opcode: READ (15), nodeid: 20, insize: 80, pid: 2101
read[0] 24576 bytes from 262144 flags: 0x8000
   read[0] 16384 bytes from 262144
   unique: 531, success, outsize: 16400
unique: 532, opcode: GETATTR (3), nodeid: 20, insize: 56, pid: 2101
getattr /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   unique: 532, success, outsize: 120
unique: 533, opcode: GETATTR (3), nodeid: 20, insize: 56, pid: 2101
getattr /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   unique: 533, success, outsize: 120
unique: 534, opcode: READ (15), nodeid: 20, insize: 80, pid: 2101
read[0] 8192 bytes from 278528 flags: 0x8000
   read[0] 8192 bytes from 278528
   unique: 534, success, outsize: 8208
unique: 535, opcode: READ (15), nodeid: 20, insize: 80, pid: 2101
read[0] 4096 bytes from 286720 flags: 0x8000
   read[0] 4096 bytes from 286720
   unique: 535, success, outsize: 4112
unique: 536, opcode: GETATTR (3), nodeid: 20, insize: 56, pid: 2101
getattr /Content/E00005538DC276AE/454109C3/00000001/Settings 20051122143332
   unique: 536, success, outsize: 120
…
```

as you can see, there are several `getattr` during the file reading : why ?

and above all, it seems that there is an additional `read`, the last one which requires 4096 bytes again, even though the entire file has already been transferred with the previous reads (`131072 + 131072 +  24576 + 8192 = 294912 > 286720`) : why ?

**Comment by N-Coder** [comment]
The commands issued depend a lot on what which program is doing on your filesystem. You already found out that `cat` uses one single read for smaller files and multiple chunked reads for bigger files, _probably checking the file size from time to time for changes using `stat`_. If you use `tail -f`, `less`, `vim` or `gedit` the order and number of requests issued may again look completely different.
For example you could do the following in python:
```
st = os.stat(file)
with open(file, 'rt') as fd:
    while 1:
        buf = fd.read(4 * 1024)
        if not buf:
            break
        print(buf)
```
which will probably lead to a sequence of stat, open, multiple reads and close, similar to the `cat` behaviour (except for intermediate `stat` calls). After the `read` calls reach the end of the file, _one final read call will be issued_, which doesn't return any data indicating the end of the file and thus `break`ing the loop.
Using `fd.seek` you could also jump to an arbitrary position in the file and read an arbitrary amount of bytes from there.

In a nutshell, don't make any assumptions about the circumstances under which the FUSE operations are invoked by other applications.
If this answers your questions, please close the bug report. ;)

**Comment by patatetom** [comment]
thanks for the explanations, and for `fusepy` of course.
regards.

**Comment by terencehonles** [comment]
Thanks for answering the question @N-Coder
